### PR TITLE
Add WooCommerce Payments features

### PR DESCRIPTION
**Pull request by aprea**
This PR adds 4 features to Jurassic Ninja, see below for an explanation of each feature and example usage.

### `woocommerce-payments`

This feature installs and activates the latest [WooCommerce Payments](https://wordpress.org/plugins/woocommerce-payments) plugin from WordPress.org.

Example use: `https://jurassic.ninja/create/?woocommerce-payments`

### `woocommerce-payments-branch`

~~This feature builds, installs and activates the WooCommerce Payments plugin for a given [`woocommerce-payments`](https://github.com/Automattic/woocommerce-payments) branch.~~

~~This is useful for `woocommerce-payments` code review as the pull request reviewer can quickly spin up a fresh JN site and test the proposed changes in the PR.~~

~~Example use: `https://jurassic.ninja/create/?woocommerce-payments-branch=try/jurassic-ninja`~~

**Update:** This feature was removed due to concerns about the impact on server resources and lack of a systems team maintaining it.

### `woocommerce-payments-dev-tools`

This feature installs and activates the trunk version of [`woocommerce-payments-dev-tools`](https://github.com/Automattic/woocommerce-payments-dev-tools) from GitHub.

This plugin is used alongside WooCommerce Payment for testing and development purposes.

Example use: `https://jurassic.ninja/create/?woocommerce-payments-dev-tools`

### `woocommerce-payments-jn-options`

This feature adds a set of WordPress options to the site to make it easier to work with WooCommerce Payments on Jurassic Ninja sites.

In particular, it disables the WPCOM request proxy and enables the UPE (Universal Payment Element) and subscriptions features.

Example use: `https://jurassic.ninja/create/?woocommerce-payments-jn-options`

### `woocommerce-payments-release`

This feature installs and activates the given tagged version of [`woocommerce-payments`](https://github.com/Automattic/woocommerce-payments) from GitHub.

This could be useful when we prepare testing instruction (e.g. pbPjnJ-xv-p2) for our testing partners, GlobalStep. This would allow us to provide a JN link to GlobalStep to install the specific WooCommerce Payments testing package for a given testing cycle.

It could also be useful for engineers/HEs when testing regressions. e.g. If we believe that a regression was introduced in the latest WooCommerce Payments release, we could install the prior release to confirm/deny this.

Example use: `https://jurassic.ninja/create/?woocommerce-payments-release=3.4.0`

### CODE REVIEW

File: features/woocommerce-payments.php
Comment (aprea) [inline features/woocommerce-payments.php:89]: **Note:** I will move this script to the `woocommerce-payments` repo when/if this PR is approved.

Comment (brbrr) [reply features/woocommerce-payments.php:89]: Historically, we have not built anything on JN server. I think there are multiple reasons why compiling anything on a production is not the best idea. At the very least, it will affect the performance and stability. 

JN is disk space bound. CPU spikes are ~30% while memory usage is stable ~20%. So I guess it would be capable of building multiple woo-payments at the same time without consequences. I'm up to do the experiment.

Comment (brbrr) [reply features/woocommerce-payments.php:89]: FWIW Jetpack's live branches are backed by the repository of pre-built branches. with some recent changes, the repository is capable to build all the plugins in Jetpack's monorepo. Maybe it could be extended to support woo universe as well?

Comment (aprea) [reply features/woocommerce-payments.php:89]: Hey @brbrr, thanks for the review! 🙏

>Historically, we have not built anything on JN server. I think there are multiple reasons why compiling anything on a production is not the best idea. At the very least, it will affect the performance and stability.

Ah, thanks for the context. FWIW this script is largely based on the existing `gutenberg-master` feature,  i.e. https://github.com/Automattic/jurassic.ninja/blob/master/features/gutenberg-master.php#L61

Apologies for my ignorance, but when you say "we have not built anything on JN server", are you referring to the build happening on the DigitalOcean droplet that the jurassic.ninja website lives on? Or are you referring to the individual Jurassic Ninja site that is deployed by ServerPilot? (or are these both hosted on the same server?)

>FWIW Jetpack's live branches are backed by the repository of pre-built branches. with some recent changes, the repository is capable to build all the plugins in Jetpack's monorepo. Maybe it could be extended to support woo universe as well?

Ah cool, that's definitely something we could explore if this approach doesn't work out. I'm definitely curious to see how that works.

One downside to my current approach is that the build takes ~6 minutes when using the `woocommerce-payments-branch` feature. This is because there are quite a few things happening in that script, i.e.
- Download and install `nvm`.
- Download and install `node`.
- Install PHP dependencies (via `composer`).
- Install JS dependencies (via `npm`).

Comment (brbrr) [reply features/woocommerce-payments.php:89]: > FWIW this script is largely based on the existing gutenberg-master feature

🙈 Ouch, I have not realized that we have that :) I just double-checked, and it seems that the feature/script is broken. So I doubt anyone is using it. 

> build happening on the DigitalOcean droplet that the jurassic.ninja website lives on
It seems that the commands are running on the server-pilot instance. So the building process should not affect the JN instance. 

> build takes ~6 minutes

Yeah, it seems a lot. with this waiting time, it would probably be faster to checkout out the PR locally.

**Comment by haszari** [comment]
Amazing! Thanks for kicking this off @aprea !

File: features/woocommerce-payments.php
Comment (aprea) [reply features/woocommerce-payments.php:89]: >Yeah, it seems a lot.

It's certainly not ideal but we could potentially decrease this build time with optimizations in the future. In any case, I don't see this being a deal-breaker for the intended use case.

The PR reviewer could request a JN site with the `woocommerce-payments-branch` feature and start reviewing the code while the build is happening.

>with this waiting time, it would probably be faster to checkout out the PR locally.

Speed was not the only motivation for this feature. Our local development environments are often not the best place to test pull requests. This could be for a number of reasons:

- The PR reviewer might already be working on their own feature branch and would need to stash/commit their changes to checkout the PR branch.
- The PR reviewer's local environment might not be in an ideal state for PR testing, i.e. they have some outdated data in their database, they have outdated local dependencies, etc.
- The PR might require the WooCommerce store to be onboarded in a specific country and the PR reviewer's local dev environment WC store is onboarded in a different country. In this scenario, it's a little disruptive to reonboard the store for PR testing and reonbard back to the original store location after completing the PR testing.
- Similar to above, the PR might introduce a change to the onboarding process and would require the PR reviewer to reonboard their local store.

@brbrr with this in mind, do you have concerns or blockers that would prevent us from moving forward with this PR?

Comment (brbrr) [reply features/woocommerce-payments.php:89]: I probably should have been clearer. I don't think any "concerns" that I mentioned are the blockers. Since it is quite an isolated feature, and it should not affect other JN users, I'm fine with it. 

FWIW I'd wait for @oskosk's feedback, since he is the creator of JN, and have a better understanding of possible drawbacks.

Comment (oskosk) [reply features/woocommerce-payments.php:89]: implementing Gutenberg in Calypso and needed an impartial build of Gutenberg  master in a self hosted to double check bugs. It was always slow to the point of being useless eventually

File: features/woocommerce-payments.php
Comment (brucealdridge) [inline features/woocommerce-payments.php:53]:
```suggestion
					$features['woocommerce-payments-dev-tools'] = true;
					$features['woocommerce'] = true;
```

File: features/woocommerce-payments.php
Comment (brucealdridge) [inline features/woocommerce-payments.php:61]:
```suggestion
					$features['woocommerce-payments-release'] = $json_params['woocommerce-payments-release'];
					$features['woocommerce'] = true;
```
It would be handy to auto-install WC if WC Payments is requested.

**Comment by aprea** [comment]
Happy new year @oskosk! 🎉 Just wondering if you have time to review this PR.

**Review by brbrr — APPROVED** [review]
I say let's try it!

**Comment by aprea** [comment]
Woo, thanks @brbrr! I'll merge this now 😄 Could you ping me when this is deployed? 🙏

**Pull request merged**
